File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    seen := make(map[rune]int)
    maxlen := 0
    i := 0

    for j, letter := range s {
        if result, ok := seen[letter]; ok && result >= i {
            i = result + 1
        }
        seen[letter] = j
        length := j - i + 1
        if length > maxlen {
            maxlen = length
        }
    }
    return maxlen
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
# longest ss containing only unique chars
# dynamic subwindow
# track max_len
# track exist in a dict

# for right in range(len(s))
    # exist[right] = 1
    # while invalid:
        # decrement left
    # max_len = max(...)
# return max_len

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_len = 0
        exist = dict()
        left = 0

        for right in range(len(s)):
            while s[right] in exist:
                exist[s[left]] -= 1
                del exist[s[left]]
                left += 1
            exist[s[right]] = 1
            max_len = max(max_len, right-left+1)
        
        return max_len